### ISSUE TITLE
2.3-Calculate Gradient by python

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#9]
在開始使用 Framework 之前，我們還是得先看看有沒有辦法利用純手工的方式，先來解決一些基本的問題，那這樣也更有助於我們去理解這些機器學習/深度學習的運作（當然重點是你會更加理解 Framework 有多好用...）

Gradient 其實就是偏微分的值組成的 vector
![image](https://user-images.githubusercontent.com/34474924/211759326-a3988db5-b12e-4d28-9c2b-14379847f252.png)

![image](https://user-images.githubusercontent.com/34474924/211759360-ae317048-ca07-4f04-bea5-7d3a9847f4d8.png)

loss function 是是用來判斷現在選擇的這個 function 有多差

偏微分是對我們的 Loss function 對變化參數（w, b, ...）去做偏微分
因此在計算 Gradient 之前，我們必須先知道我們的 Loss function 是什麼

假設我們的 Loss function 是 mean square error（均方誤差）
![image](https://user-images.githubusercontent.com/34474924/211762976-9907d8f5-253b-4c6d-85c2-5ee6ca4f5175.png)

```python
import numpy as np

def loss(y, y_predicted):

    return ((y_predicted-y) ** 2).mean()
```

那我們還需要知道總共有哪些變數，因此我們假設今天我們是在算 Regression

function 就是
![image](https://user-images.githubusercontent.com/34474924/211764057-7ee3c496-4be6-4d5e-b56a-3ac8b04bdc90.png)
變數就是一個 w

所以 MSE 也會寫成 
![image](https://user-images.githubusercontent.com/34474924/211764154-c5b21811-d0ba-43bc-8234-60bdeb86e738.png)

_______________________
### Gradient

![image](https://user-images.githubusercontent.com/34474924/211764538-cdcfde3c-3c70-455e-93bc-3079739598dc.png)

```python
def gradient(x, y, y_predicted):

    return np.dot(2*x, y_predicted - y).mean()
```

上面的流程也就是 Model -> Loss function -> Gradient Descent -> Find the best function

我們先假定我們的目標函數是 y=2*x

那在不知道答案的情況下，我們可以先假設 Model 會長y=w*x, 我們今天要解決的問題就是找出正確的 w

```py
import numpy as np

# f = w * x
# f = 2 * x, we set w as 2

#人為生成一些資料, x和y是進行ML之前就需要具備的資料
x = np.array([1, 2, 3, 4, 5, 6], dtype=np.float32)
y = np.array([2, 4, 6, 8, 10, 12], dtype=np.float32)

# init weight 一開始是y=0.0*w
w = 0.0

# model prediction
# 這邊之後會解釋為啥叫做 forward，可以先視為計算函數而已
def forward(x):
    
    return w * x
    
# set up loss function as mean square error
def loss(y, y_predicted):

    return ((y_predicted-y) ** 2).mean()
    
# gradient
def gradient(x, y, y_predicted):

    return np.dot(2*x, y_predicted-y).mean()
    
print(f'Prediction before training: f(5) = {forward(5): .3f}')

# Training
learning_rate = 0.01
n_iters = 10

for epoch in range(n_iters):
    # perdiction = forward pass
    y_pred = forward(x)

    # loss
    l = loss(y, y_pred)

    # we know that gradient descent is where 
    # calculate gradient and update parameters
    
    # calculation of gradients
    dw = gradient(x, y, y_pred)

    # update weights
    w -= learning_rate * dw

    print(f'epoch {epoch + 1}: w = {w:.3f}, loss = {l:.8f}')

print(f'Prediction after training: f(5) = {forward(5): .3f}')
```

[https://ithelp.ithome.com.tw/articles/10267426](https://ithelp.ithome.com.tw/articles/10267426)
